tooling/bash-limits.md

```diff
@@ -0,0 +1,228 @@
+1. **Extremely Large Datasets**
+	- BASH does not have memory management or garbage collection.
+	- Variables and arguments have a size limit based on the memory of a machine, which is not necessarily deterministic.
+	- BASH utilities have their own limits. For example, `sed` on BSD and macOS can only operate on files of 4 MB or less.
+	- Working with large datasets in BASH can cause scripts to become extremely slow.
+	- "Real" languages allow an engineer to predict these limits and performance on large datasets much more accurately, even independent of a specific host machine's hardware.
+1. **Concurrency** - use a "real" language for multithreading or parallelization.
+1. **Recursion** - functions which call themselves.
+	```bash
+	function factorial {
+		if [[ $1 -eq 1 ]]; then
+			echo 1
+		else
+			local result=$(( $1 * $(factorial $(( $1 - 1 ))) ))
+			echo $result
+		fi
+	}
+
+	factorial 5  # 120
+	```
+1. **Higher-Order Functions** - functions which take a function as an argument or return a function as a result.
+	- In BASH, you would be passing around names of functions.
+	```bash
+	function add {
+		echo $(( $1 + $2 ))
+	}
+
+	function multiply {
+		echo $(( $1 * $2 ))
+	}
+
+	function calculate {
+		local operation=$1
+		shift
+		$operation $@
+	}
+
+	calculate add 5 3  # 8
+	calculate multiply 5 3  # 15
+	```
+1. **Currying** - where your function has completely different behavior depending on how many arguments are passed, a type of polymorphism.
+	- Example in Haskell.
+		```haskell
+		add :: Int -> Int -> Int
+		add x y = x + y
+
+		add5 :: Int -> Int
+		add5 = add 5
+
+		main = print (add5 3)  -- 8
+		```
+	- Example in JavaScript.
+		```js
+		const add = x => y => x + y;
+		const add5 = add(5);
+		console.log(add5(3));  // 8
+		```
+	- Example in Python.
+		```python
+		from functools import partial
+
+		def add(a, b):
+			return a + b
+
+		add_5 = partial(add, 5)
+
+		print(add_5(3))  # 8
+		```
+	- Another example in Pythin using closure and higher-order functions instead of partial application.
+		```python
+		def curry(f):
+			def curried(*args, **kwargs):
+				if len(args) + len(kwargs) >= f.__code__.co_argcount:
+					return f(*args, **kwargs)
+				return lambda *args2, **kwargs2: curried(*(args + args2), **{**kwargs, **kwargs2})
+			return curried
+
+		@curry
+		def add(x, y):
+			return x + y
+
+		add5 = add(5)
+
+		print(add5(3))  # 8
+		```
+	- Example attempting this in BASH.
+		```bash
+		function add {
+			local x=$1
+			local y=$2
+			echo $((x + y))
+		}
+
+		function curried_add {
+			local x=$1
+			if [[ $# -eq 1 ]]; then
+				echo "add $x \$1"
+			else
+				echo $(add $x $2)
+			fi
+		}
+
+		add_5=$(curried_add 5)
+		$add_5 3  # 8
+		```
+1. **Partial Application** - where a function is called with fewer arguments than it is defined to take, resulting in a new function that takes the remaining arguments.
```

Comment:
Because it's obviously acceptable and correct to have default arguments, this is either wrong or I don't understand what you're getting at. Maybe this is unnecessary and distracting?

Reply:
Partial application is not the use of default arguments for parameters. Partial application is a functional programming paradigm. While it can be approximated in non-functional languages, examples in non-functional languages may not make sense. Please review the example of partial application in Haskell, and I suspect you will agree we should not be trying to do this in BASH scripts.

Reply:
I see you marked this conversation resolved, but I "unresolved" it because I _do not_ feel my previous comment was an acceptable response to peer review feedback and I owe you an apology. I was in a meeting at the time and should have given this feedback more attention and consideration. This specific list element has caused some confusion and it deserves clarification that anyone can understand, not just Haskell programmers.

I have added a note to the "partial application" list element providing a more detailed explanation of partial application, and clarifying this list element does **_not_** refer to using default values for function parameters.

Please read the [updated version of the document](https://github.com/eosnetworkfoundation/engineering/blob/968164405acb780f3117e5219934cb0c5d8acd3a/tooling/bash-limits.md#red-flags) and let me know if that clarifies the concept or not.

_After_ you have read the document to help me determine if it communicates what I am trying to communicate, you can check out these other explanations of partial application that I did not include in the document, if you think that would be helpful.

<details>
<summary>See More</summary>
These are pretty much all copy/pasta from my conversation with ChatGPT trying to understand the concept thoroughly myself, especially how it is distinct from other functional programming concepts because they are often used together, then generating code samples for the document.

> Partial application is a technique in functional programming where a function that takes multiple arguments is partially applied with some arguments pre-specified, resulting in a new function that takes fewer arguments. This allows for the creation of new functions with some arguments pre-specified, which can be useful for creating more flexible and reusable code.

Here is an example it gave me in Python which is simpler than the one in the document, but which I excluded because it uses a library instead of "pure" or "native" code if you know what I mean.

> Sure, here's an example of partial application in Python, similar to the one I showed you in JavaScript:
> ```python
> from functools import partial
>
> def multiply(a, b):
>     return a * b
>
> double = partial(multiply, 2)
>
> print(double(3)) # 6
> ```
> In this example, we have a function `multiply` that takes two arguments, `a` and `b`, and returns their product. We then use the `functools.partial` function to create a new function `double` which is a partially applied version of the `multiply` function. The `functools.partial` function takes the original function as the first argument and then any number of arguments to fix in place. In this example, we're fixing the first argument of the `multiply` function to be 2, so the double function will always multiply its argument by 2.
>
> Now when we call `double(3)` it will return the result of `multiply(2,3)` which is 6.

Here is one of the Python examples I used in the document with the prose explanation from ChatGPT.

> Sure, here's an example of partial application in Python that is similar to the example in JavaScript:
> ```python
> def add(x, y):
>     return x + y
>
> add5 = lambda y: add(5, y)
>
> print(add5(3)) # 8
> ```
> In this example, we have a function `add` that takes two arguments, `x` and `y`, and returns their sum. Then we use lambda function to create a new function `add5` which is a partially applied version of the `add` function. The lambda function takes one argument `y` and returns the result of `add(5, y)`.
>
> Now when we call `add5(3)` it will return the result of `add(5,3)` which is `8`.

Here is the prose explanation of the Haskell example included in the document, if you are curious.

> Sure, here's an example of partial application in Haskell that is similar to the example in JavaScript:
> ```haskell
> add :: Int -> Int -> Int
> add x y = x + y
>
> add5 :: Int -> Int
> add5 = add 5
>
> main = print (add5 3)
> ```
> In this example, we have a function `add` that takes two arguments, `x` and `y`, and returns their sum. Then we use the function `add` to create a new function `add5` which is a partially applied version of the `add` function. The `add5` function takes one argument and returns the result of `add 5` and the passed argument.
>
> Now when we call `add5 3` it will return the result of `add 5 3` which is 8.
>
> It's important to note that in Haskell, when defining a function, the number of parameters in the function signature defines how many arguments need to be passed to the function to get a value. In the example above, `add :: Int -> Int -> Int` means that `add` is a function that takes two `Int` arguments and returns an `Int` value.

Yet another explanation. ChatGPT likes to repeat itself if it has extra space in the output.

> Partial application, on the other hand, is a functional programming concept that refers to the creation of a new function by fixing some of the arguments to a function and leaving others unfixed. It allows you to create a new function with fewer parameters that is easier to call, as well as providing more control over the arguments passed to the original function. The example in the JavaScript code you showed earlier is an example of partial application.

Hopefully these various descriptions and additional examples are helpful. I know it helps me to hear the same thing multiple times or stated multiple different ways to understand a new concept, or to reinforce my understanding of an existing concept.

</details>

Reply:
I like the change you made. I feel like a Junior can sort these now.